src/lib/parser.c

```diff
@@ -715,27 +721,27 @@ static inline RdbStatus unpackList(RdbParser *p, unsigned char *lp) {
              * 2. allocFromCache(RQ_ALLOC_APP_BULK_REF) will:
              *    - Set last character '\0' to terminate the string.
              *    - Mark the string as a referenced bulk allocation (placement-new alloc)
+             *      Note: IF app expects APP_BULK, then it is not possible to return
+             *            a reference. A new memory will be allocated instead and gets a copy
              * 3. invoke CALL_HANDLERS_CB that will:
              *    - Supply the bulk to callbacks
              *    - Finalize by restoring original char (from tmp) */
-            tmp = item[vlen];
-            IF_NOT_OK_RETURN(allocFromCache(p, vlen, RQ_ALLOC_APP_BULK_REF, item, &binfo));
-
-            /* if requested ref another memory but forced to allocate a new buffer,
-             * (since configured RDB_BULK_ALLOC_EXTERN) then copy data to the new buffer */
-            if (binfo->bulkType != BULK_TYPE_REF)
-                memcpy(binfo->ref, item, vlen);
+            tmp = *(itemEnd = itemStart + vlen);
 
+            IF_NOT_OK_RETURN(allocFromCache(p, vlen, RQ_ALLOC_APP_BULK_REF, itemStart, &binfo));
```

Comment:
Here is the comment that is written above this line (but doesn't appear in the change):
```
            /* The callback function expects a native string that is terminated
             * with '\0'. However, the string we have is packed without
             * termination. To avoid allocating a new string, we can follow these
             * steps:
             * 1. Save the last character that comes at the end of the referenced
             *    string in a temporary char (tmp).
             * 2. allocFromCache(RQ_ALLOC_APP_BULK_REF) will:
             *    - Set last character '\0' to terminate the string.
             *    - Mark the string as a referenced bulk allocation (placement-new alloc)
             *      Note: IF app expects APP_BULK, then it is not possible to return
             *            a reference. A new memory will be allocated instead and gets a copy
             * 3. invoke CALL_HANDLERS_CB that will:
             *    - Supply the bulk to callbacks
             *    - Finalize by restoring original char (from tmp) */
```
Please lmk if it is clear enough.

Reply:
i did see the comment, and i was trying to understand what you changed and why.
now that i look at it again, i have another concern.
you're not doing a backup / restore and override of the last chat with `\0`, since that will corrupt the string (one byte will be missing).
what you're doing is setting the byte after the last to `\0`, but that could in theory be beyond the size of the allocation.
however, since it's a listpack, we're guaranteed to have a terminating byte at the end of the listpack, so that makes it safe.
i think you should improve the comment with that info.